FORMAT: 1A

# Plant Breeding API
The Breeding API specifies a standard interface for plant phenotype/genotype databases to serve
their data to crop breeding applications. It is a <i>shared</i>, <i>open</i> API, to be
used by all data providers and data consumers who wish to participate. Initiated in May 2014, it is
currently in an actively developing state, so now is the time for potential participants to help
shape the specifications to ensure their needs are addressed. The listserve for discussions and
announcements is at http://mail2.sgn.cornell.edu/cgi-bin/mailman/listinfo/plant-breeding-api .
Additional documentation is in the <a href="https://github.com/plantbreeding/documentation">Github wiki</a>.

### URL structure

API requests are structured as "\<server\>/brapi/v1/",
where "v1" is the version number of the API, followed by the command.
Example: /brapi/v1/markerprofiles/2939

To distinguish between multiple databases or crops available from the same server, include the database or crop name as part of the "\<server\>" identifier.

Example: superBreedingServer.org/maize1/brapi/v1/markerprofiles/2939

### Structure of the response object:
The return objects are encoded in JSON.
The response always consists of a "metadata" key that minimally
contains the pagination information in a "pagination" key and the status information as well as a "datafiles" key that lists URLs to data files generated by the call.

If the response is a single record that doesn't require pagination, then the value for the "pagination" key is an empty hash.  When the results are paginated,
the pagination object contains the keys "pageSize", "currentPage", "totalCount", "totalPages". The first page will be page 0 (zero).

The status object contains a list of objects with the keys "code" and "message".

The payload data is contained in a key called "result". This should usually contain a list; pagination will happen over items in this list.

Additional documentation is in the [GitHub wiki](https://github.com/plantbreeding/documentation/wiki).
See especially the [Best Practices and Conventions]
(https://github.com/plantbreeding/documentation/wiki/Best-Practices-and-Conventions).

### API call categories:
Scope: "CORE", "PHENOTYPING", "GENOTYPING", "OTHER".
Status: "STABLE", "ACCEPTED", "IN DISCUSSION", "SUGGESTED".

# Group Authentication
This resource refers to the authentication mechanism for the API. It is still implementation-agnostic but the structure
should be similar across all implementations. To start off, we propose to use a call similar to OAuth2.0 password grant type. However, this is not fully secure, as we have to trust third party applications not to do anything nefarious with the password information. Upgrade to full OAuth2.0 in the future.
To log in, we use a POST call and to log out we use a DELETE call. In order to remain as compliant as possible with OAuth2 ( https://tools.ietf.org/html/rfc6749 , 7.1.  Access Token Types)
the access token should be provided by the client in its HTTP header using the format "Authorization: Bearer {bearer_value}".
When authenticated, all the calls, including the authentication call, should be done through HTTPS in order to keep the token safe.
The token life time can be "refreshed" after each call but this approach is not OAuth2 compliant.

**HTTPS** should be enforced everywhere.

## Authentication [/brapi/v1/token?grant_type=password&username=USERNAME&password=PASSWORD&client_id=CLIENT_ID]

Status: ACCEPTED.
Implementation target date: PAG2016
Implemented by: Tripal Brapi module, Cassavabase

###### Response data types
returns a hash with the session name as key and the token as the value. A metadata key is also present
(but usually set to null, unless an error condition occurs).

### Login [POST]
+ Parameters
    + grant_type (optional, text, `password`) ... The grant type, only allowed value is password, but can be ignored
    + username (required, text, `thepoweruser`) ... The username
    + password (optional, text, `mylittlesecret`) ... The password
    + client_id (optional, text, `blabla`) ... The client id, currently ignored.

+ Response 201 (application/json)

        {
            "metadata": {
                "pagination" : {},
                "status" : [],
                "datafiles": []
                },
            "userDisplayName": "John Smith",
            "access_token": "R6gKDBRxM4HLj6eGi4u5HkQjYoIBTPfvtZzUD8TUzg4",
            "expires_in": "The lifetime in seconds of the access token"
        }


## Log off [/brapi/v1/token?access_token=TOKEN]

Status: ACCEPTED.
Implementation target date: PAG2016
Implemented by: Tripal Brapi module, Cassavabase

###### Response data types
returns a hash with the session name as key and the token as the value. A metadata key is also present
(but usually set to null, unless an error condition occurs).

### Logout [DELETE]

+ Parameters
    + access_token (optional, text, `R6gKDBRxM4HLj6eGi4u5HkQjYoIBTPfvtZzUD8TUzg4`) ... The user access token. Default: current user token.

+ Response 201 (application/json)

        {
            "metadata": {
                "pagination" : {},
                "status" : ["User has been logged out successfully."],
                "datafiles": []
                }
        }


# Group Calls

## Call Search [/brapi/v1/calls?call={callName}&datatype={dataFile}&pageSize={pageSize}&page={page}]

Implemented by: Germinate

This call returns metadata about the individual BrAPI calls. Initially, this will be:
- A list of the supported data types for file export.
- A list of the HTTP methods that are supported ("GET", "POST", etc).

##### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|datafiles|list|List of the data files generated by the call|Y|
|data|array of objects|Array (possibly empty) of call records|Y|
|call|string|The relative URL/path of the call as specified in the Apiary (e.g. "germplasm/id/mcpd")|Y|
|datatypes|list|List of the data formats supported for data export, e.g. "json", "tsv", etc.|Y|
|methods|list|List of the HTTP methods supported by this call|Y|

### Call search [GET]
+ Parameters
    + call (optional, string, `allelematrix`) ... The relative URL/path as specified on Apiary
    + datatype (optional, string, `tsv`) ... The data type supported by the call
    + pageSize (optional, integer, `1000`) ... The size of the pages to be returned. Default is `1000`.
    + page (optional, integer, `10`) ... Which result page is requested
+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 3,
                    "currentPage": 0,
                    "totalCount": 3,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data": [
                    { "call": "allelematrix", "datatypes": ["json", "tsv"], "methods": ["GET", "POST"] },
                    { "call": "germplasm/id/mcpd", "datatypes": ["json"], "methods": ["GET"] },
                    { "call": "doesntexistyet", "datatypes": ["png", "jpg"], "methods": ["GET"] }
                ]
            }
        }

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 1,
                    "currentPage": 0,
                    "totalCount": 1,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data": [
                    { "call": "allelematrix", "datatypes": ["json", "tsv"], "methods": ["GET", "POST"] }
                ]
            }
        }


# Group Germplasm
Note: The plural of germplasm is germplasm (no "s").


## Germplasm Search [/brapi/v1/germplasm?name={name}&matchMethod={matchMethod}&include={synonyms}&pageSize={pageSize}&page={page}]
Scope: CORE.  Status: STABLE.
Implementation target date: PAG2016
Implemented by: Germinate, Flapjack, Tripal Brapi Client, Cassavabase

Other service requests use the server's internal `germplasmDbId`. This service returns germplasm records that provide the germplasmDbId.
For the requested name or synonym, returns an array (possibly empty) of germplasm records that match the search criteria.
- If there is none, an empty array is returned.
- If there is one or more than one match, returns an array of all germplasm records that match the search criteria.

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|result|object|data|Y|
|data|array of objects|Array (possibly empty) of germplasm records|Y|
|germplasmDbId|string|Internal db identifier|Y|
|defaultDisplayName|string|A string representing the germplasm that will be meaningful to the user|Y|
|germplasmName|string|Name of the germplasm. It can be the prefered name and does not have to be unique||
|accessionNumber|string|This is the unique identifier for accessions within a genebank, and is assigned when a sample is entered into the genebank collection||
|germplasmPUI|string|Permanent identifier (e.g. URI, DOI, LSID)||
|pedigree|string|Cross name with optional selection history.||
|seedSource|string|Seed source||
|synonyms|array of string|List of other names for this germplasm||


### Search names to retrieve germplasm records [GET]
+ Parameters
    + name (optional, text, `morex mor% mo* mo?ex`) ... The name or synonym.
    + matchMethod (optional, string, `wildcard`) ... Possible values are 'case_insensitive', 'exact' (case sensitive), 'wildcard'
    (which is case insensitive). Wildcard uses both '*' and '%' for any number of characters and '?' for one-character matching.
    Default is exact. Note: for implementation be aware that security issues (e.g. sql injection) are the responsibility of the developer.
    + include (optional, string, `synonyms`) ... Whether to include synonyms in the output.
    + pageSize (optional, integer, `1000`) ... The size of the pages to be returned. Default is `1000`.
    + page (optional, integer, `10`) ... Which result page is requested
+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 1000,
                    "currentPage": 10,
                    "totalCount": 27338,
                    "totalPages": 28
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data": [
                    {
                        "germplasmDbId": "382",
                        "defaultDisplayName": "Pahang",
                        "germplasmName": "Pahang",
                        "accessionNumber": "ITC0609",
                        "germplasmPUI": "http://www.crop-diversity.org/mgis/accession/01BEL084609",
                        "pedigree": "TOBA97/SW90.1057",
                        "seedSource": "",
                        "synonyms": ["01BEL084609"],
                    },
                    {
                        "germplasmDbId": "394",
                        "defaultDisplayName": "Pahang",
                        "germplasmName": "Pahang",
                        "accessionNumber": "ITC0727",
                        "germplasmPUI": "http://www.crop-diversity.org/mgis/accession/01BEL084727",
                        "pedigree": "TOBA97/SW90.1057",
                        "seedSource": "",
                        "synonyms": [ "01BEL084727"],
                    }
                ]
            }
        }



##  Germplasm MCPD Search [/brapi/v1/germplasm/mcpd?germplasmName=&germplasmGenus=&germplasmSubTaxa=&germplasmDbId&germplasmPUI=http://data.inra.fr/accession/234Col342&germplasmSpecies=Triticum&panel=diversitypanel1&collection=none&pageSize=pageSize&page=page]
Scope: CORE. Status: IN DISCUSSION.
Implementation target date: PAG2017
Implemented by: GnpIS

Adresses two needs :
1. possibility to search germplams by more parameters than those alowed by the existing germplasm search
2. possibility to get MCPD details by PUID rather than dbId

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|result|object|Object containing MCPD data|Y|
|germplasmDbId|string|Internal db identifier|Y|
|defaultDisplayName|string|A string that can be displayed to the user|Y|
|accessionNumber|string|This is the unique identifier for accessions within a genebank, and is assigned when a sample is entered into the genebank collection||
|germplasmName|string|Name of the germplasm. It can be the prefered name and does not have to be unique||
|germplasmPUI|string|Permanent identifier (e.g. URI, DOI, LSID)||
|pedigree|string|Cross name with optional selection history.||
|seedSource|string|Seed source||
|synonyms|array of string|List of other germplasm name||
|commonCropName|string|Common name fo the crop (e.g. wheat, rice, maize, cassava, banana)||
|instituteCode|string|Institute that has bred the material. Note: The code may consist of the 3-letter ISO 3166 country code of the country where the institute is located plus a number (e.g. COL001) as recommended by FAO WIEWS |Y|
|instituteName|string|Name of the institute (or person) that bred the material.||
|biologicalStatusOfAccessionCode|string|400) Breeding/research material 410) Breeder's line 411) Synthetic population 412) Hybrid 413) Founder stock/base population 414) Inbred line (parent of hybrid cultivar) 415) Segregating population 416) Clonal selection 420) Genetic stock 421) Mutant (e.g. induced/insertion mutants, tilling populations) 422) Cytogenetic stocks  423) Other genetic stocks (e.g. mapping populations)500) Advanced or improved cultivar (conventional breeding methods) 600) GMO (by genetic engineering) 999) Other||
|countryOfOriginCode|string|3-letter ISO 3166-1 code of the country in which the sample was bred or selected (breeding lines, GMOs, segregating populations, hybrids, modern cultivars, etc.).||
|typeOfGermplasmStorageCode|array of string|If germplasm is maintained under different types of storage, multiple choices are allowed. 10) Seed collection 11) Short term 12) Medium term 13) Long term 20) Field collection 30) In vitro collection 40) Cryopreserved collection 50) DNA collection 99) Other (elaborate in REMARKS field)||
|genus|string|Genus name for taxon. Initial uppercase letter required.||
|species|string|Specific epithet portion of the scientific name in lowercase letters.||
|speciesAuthority|string|||
|subtaxa|string|Subtaxon can be used to store any additional taxonomic identifier. The following abbreviations are allowed: ‘subsp.’ (for subspecies); ‘convar.’ (for convariety); ‘var.’ (for variety); ‘f.’ (for form); ‘Group’ (for ‘cultivar group’).|
|subtaxaAuthority|string|||
|donors|array of object|code of the donor institute and Identifier assigned to an accession by the donor, and permanent identifier.||
|acquisitionDate|string|Date on which the accession entered the collection where YYYY is the year, MM is the month and DD is the day. Missing data (MM or DD) should be indicated with hyphens or ‘00’ [double zero].|

### Germplasm MCPD search [GET]
+ Parameters
    + germplasmPUI (optional, text, `http://data.inra.fr/accession/234Col342`) ... The name or synonym of external genebank accession identifier
    + germplasmDbId (optional, text, `986`) ... The name or synonym of external genebank accession identifier
    + germplasmSpecies (optional, text, `aestivum`) ... The name or synonym of genus or species ( merge with below ?)
    + germplasmGenus (optional, text, `Triticum, Hordeum`) ... The name or synonym of genus or species
    + germplasmName (optional, text, `Triticum, Hordeum`) ... The name or synonym of the accession
    + germplasmSubTaxa (optional, text, `cv. Charger, subsp. aestivum`) ... The name or synonym of MCPD subTaxa. Exact Match, abreviations must be MCPD compliant (‘subsp.’ for 'subspecies'; ‘convar.’ for "convariety" ‘var.’ for variety; ‘f.’ for 'form'; ‘Group’ for ‘cultivar group’)
    + panel (optional, text, `breedingProgramPanel2011`) ... The name of a specific panel
    + collection (optional, text, `BRCCollection_Wheat`) ... The name of a specific Collection
    + pageSize (optional, integer, `1000`) ... The size of the pages to be returned. Default is `1000`.
    + page (optional, integer, `10`) ... Which result page is requested

+ Response 200 (application/json)

        {
            "metadata": {
                "status": [],
                "datafiles": [],
                "pagination": {
                "pageSize": 10,
                "currentPage": 1,
                "totalCount": 2,
                "totalPages": 1
                }
            },
            "result": {
                "data":[
                    {
                        "germplasmDbId": "01BEL084609",
                        "defaultDisplayName": "Pahang",
                        "accessionNumber": "ITC0609",
                        "germplasmName": "Pahang",
                        "germplasmPUI": "http://www.crop-diversity.org/mgis/accession/01BEL084609",
                        "pedigree": "TOBA97/SW90.1057",
                        "germplasmSeedSource": "Female GID:4/Male GID:4",
                        "synonyms": [ ],
                        "commonCropName": "banana",
                        "instituteCode": "01BEL084",
                        "instituteName": "ITC",
                        "biologicalStatusOfAccessionCode": 412,
                        "countryOfOriginCode": "UNK",
                        "typeOfGermplasmStorageCode": 10,
                        "genus": "Musa",
                        "species": "acuminata",
                        "speciesAuthority": "",
                        "subtaxa": "sp malaccensis var pahang",
                        "subtaxaAuthority": "",
                        "donors":
                        [
                            {
                                "donorAccessionNumber": "",
                                "donorInstituteCode": "",
                                "germplasmPUI": ""
                            }
                        ],
                        "acquisitionDate": "19470131"
                    ,{
                        "germplasmDbId": "03REL084609",
                        "defaultDisplayName": "Pah",
                        "accessionNumber": "ITC0685",
                        "germplasmName": "Pah",
                        "germplasmPUI": "doi:10.15454/328757862534E12",
                        "pedigree": "TOBA97/SW90.1057",
                        "germplasmSeedSource": "Female GID:4/Male GID:4",
                        "synonyms": [ ],
                        "commonCropName": "banana",
                        "instituteCode": "01BEL084",
                        "instituteName": "ITC",
                        "biologicalStatusOfAccessionCode": 412,
                        "countryOfOriginCode": "UNK",
                        "typeOfGermplasmStorageCode": 10,
                        "genus": "Musa",
                        "species": "acuminata",
                        "speciesAuthority": "",
                        "subtaxa": "sp malaccensis var pah",
                        "subtaxaAuthority": "",
                        "donors":
                        [
                            {
                                "donorAccessionNumber": "",
                                "donorInstituteCode": "",
                                "germplasmPUI": ""
                            }
                        ],
                        "acquisitionDate": "19770131"
                    }
                ]
            }
        }



## Germplasm Details by germplasmDbId [/brapi/v1/germplasm/{id}]
Scope: CORE. Status: STABLE.
Implementation target date: PAG2016
Implemented by: Tripal Brapi module, Cassavabase, Germinate

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|result|object|data|Y|
|data|array of object|Array of 0 or 1 germplasm records|Y|
|germplasmDbId|string|Internal db identifier|Y|
|defaultDisplayName|string|A string representing the germplasm that will be meaningful to the user|Y|
|germplasmName|string|Name of the germplasm. It can be the prefered name and does not have to be unique||
|accessionNumber|string|This is the unique identifier for accessions within a genebank, and is assigned when a sample is entered into the genebank collection||
|germplasmPUI|string|Permanent identifier (e.g. URI, DOI, LSID)||
|pedigree|string|Cross name with optional selection history.||
|seedSource|string|Seed source||
|synonyms|array of string|List of other germplasm name||

### Germplasm Details by id [GET]
+ Parameters
    + id (required, string, `382`) ... the internal id of the germplasm

+ Response 200 (application/json)

        {
            "metadata": {
                "status": [],
                "datafiles": [],
                "pagination": {}
            },
            "result": {
                "germplasmDbId": "01BEL084609",
                "defaultDisplayName": "Pahang",
                "germplasmName": "Pahang",
                "accessionNumber": "ITC0609",
                "germplasmPUI": "http://www.crop-diversity.org/mgis/accession/01BEL084609",
                "pedigree": "TOBA97/SW90.1057",
                "seedSource": "Female GID:4/Male GID:4",
                "synonyms": ["Pahanga","Pahange"],
            }
        }

##  Germplasm Multi Crop Passport Data [/brapi/v1/germplasm/{id}/mcpd]
Scope: CORE. Status: ACCEPTED.
Implementation target date: PAG2016
Implemented by: Tripal Brapi module, Germinate

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|result|object|Object containing MCPD data|Y|
|germplasmDbId|string|Internal db identifier|Y|
|defaultDisplayName|string|A string that can be displayed to the user|Y|
|accessionNumber|string|This is the unique identifier for accessions within a genebank, and is assigned when a sample is entered into the genebank collection||
|germplasmName|string|Name of the germplasm. It can be the prefered name and does not have to be unique||
|germplasmPUI|string|Permanent identifier (e.g. URI, DOI, LSID)||
|pedigree|string|Cross name with optional selection history.||
|seedSource|string|Seed source||
|synonyms|array of string|List of other germplasm name||
|commonCropName|string|Common name fo the crop (e.g. wheat, rice, maize, cassava, banana)||
|instituteCode|string|Institute that has bred the material. Note: The code may consist of the 3-letter ISO 3166 country code of the country where the institute is located plus a number (e.g. COL001) as recommended by FAO WIEWS |Y|
|instituteName|string|Name of the institute (or person) that bred the material.||
|biologicalStatusOfAccessionCode|string|400) Breeding/research material 410) Breeder's line 411) Synthetic population 412) Hybrid 413) Founder stock/base population 414) Inbred line (parent of hybrid cultivar) 415) Segregating population 416) Clonal selection 420) Genetic stock 421) Mutant (e.g. induced/insertion mutants, tilling populations) 422) Cytogenetic stocks  423) Other genetic stocks (e.g. mapping populations)500) Advanced or improved cultivar (conventional breeding methods) 600) GMO (by genetic engineering) 999) Other||
|countryOfOriginCode|string|3-letter ISO 3166-1 code of the country in which the sample was bred or selected (breeding lines, GMOs, segregating populations, hybrids, modern cultivars, etc.).||
|typeOfGermplasmStorageCode|array of string|If germplasm is maintained under different types of storage, multiple choices are allowed. 10) Seed collection 11) Short term 12) Medium term 13) Long term 20) Field collection 30) In vitro collection 40) Cryopreserved collection 50) DNA collection 99) Other (elaborate in REMARKS field)||
|genus|string|Genus name for taxon. Initial uppercase letter required.||
|species|string|Specific epithet portion of the scientific name in lowercase letters.||
|speciesAuthority|string|||
|subtaxa|string|Subtaxon can be used to store any additional taxonomic identifier. The following abbreviations are allowed: ‘subsp.’ (for subspecies); ‘convar.’ (for convariety); ‘var.’ (for variety); ‘f.’ (for form); ‘Group’ (for ‘cultivar group’).|
|subtaxaAuthority|string|||
|donors|array of object|code of the donor institute and Identifier assigned to an accession by the donor, and permanent identifier.||
|acquisitionDate|string|Date on which the accession entered the collection where YYYY is the year, MM is the month and DD is the day. Missing data (MM or DD) should be indicated with hyphens or ‘00’ [double zero].|

### Germplasm MCPD by id [GET]
+ Parameters
    + id (required, string, `382`) ... germplasmDbId - the internal id of the germplasm

+ Response 200 (application/json)

        {
            "metadata": {
                "status": [],
                "datafiles": [],
                "pagination": {}
            },
            "result": {
                "germplasmDbId": "01BEL084609",
                "defaultDisplayName": "Pahang",
                "accessionNumber": "ITC0609",
                "germplasmName": "Pahang",
                "germplasmPUI": "http://www.crop-diversity.org/mgis/accession/01BEL084609",
                "pedigree": "TOBA97/SW90.1057",
                "germplasmSeedSource": "Female GID:4/Male GID:4",
                "synonyms": [ ],
                "commonCropName": "banana",
                "instituteCode": "01BEL084",
                "instituteName": "ITC",
                "biologicalStatusOfAccessionCode": 412,
                "countryOfOriginCode": "UNK",
                "typeOfGermplasmStorageCode": 10,
                "genus": "Musa",
                "species": "acuminata",
                "speciesAuthority": "",
                "subtaxa": "sp malaccensis var pahang",
                "subtaxaAuthority": "",
                "donors":
                [
                    {
                        "donorAccessionNumber": "",
                        "donorInstituteCode": "",
                        "germplasmPUI": ""
                    }
                ],
                "acquisitionDate": "19470131"
            }
        }

## Germplasm Details List by studyDbId [/brapi/v1/studies/{studyDbId}/germplasm?pageSize=20&page=1]
Scope: CORE.
Status: STABLE.
Implemented by: Cassavabase, Tripal Brapi Module, Germinate

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|trialDbId|string|The internal Id of the trial|Y|
|trialName|string|Name of the trial||
|data|array of objects|Each object is a germplasm record representing one of the entries in the trial|Y|
|germplasmDbId|string|Internal db identifier|Y|
|trialEntryNumberId|string|Entry number for this germplasm in the trial||
|defaultDisplayName|string|A string representing the germplasm that will be meaningful to the user|Y|
|germplasmName|string|Name of the germplasm. It can be the prefered name and does not have to be unique||
|accessionNumber|string|This is the unique identifier for accessions within a genebank, and is assigned when a sample is entered into the genebank collection||
|germplasmPUI|string|Permanent identifier (e.g. URI, DOI, LSID)||
|pedigree|string|Cross name with optional selection history.||
|seedSource|string|Seed source||
|synonyms|array of string|List of other germplasm name||


## Germplasm pedigree [/brapi/v1/germplasm/{id}/pedigree?notation=purdy]
Scope: CORE. Status: IN DISCUSSION.
Implementation target date: PAG2016
Implemented by: Germinate, Tripal Brapi Module, Cassavabase (without notation option)

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|germplasmDbId|string|Internal db identifier|Y|
|defaultDisplayName|string|A string representing the germplasm that will be meaningful to the user|Y|
|pedigree|string|Cross name with optional selection history.|Y|
|parent1Id|string|germplasmDbId of parent1||
|parent2Id|string|germplasmDbId of parent2||

(http://wheat.pw.usda.gov/ggpages/gopher/administration/Template%20for%20Germplasm%20records.html) or [Lamacraft] (http://link.springer.com/article/10.1007%2FBF00021556).
### Germplasm pedigree by id [GET]
+ Parameters
    + id (required, number, `382`) ... the internal id of the germplasm
    + notation (optional, string, `purdy`) ... text representation of the pedigree
+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {},
                "status": [],
                "datafiles": []
            }
            "result" : {
                "germplasmDbId": 382,
                "pedigree" : "Cree / Bonanza",
                "parent1Id" : 166,
                "parent2Id" : 143
            }
        }

+ Response 400 (application/json)

        {
            "metadata" : {
                "status" : [
                    {
                        "code" : "ERR-1",
                        "message" : "Unsupported notation code."
                    }
                ],
                "datafiles": []
            },
            "result": {}
        }

## Germplasm markerprofiles  [/brapi/v1/germplasm/{id}/markerprofiles]
Retrieve the markerProfileDbIds for a given Germplasm ID
Scope: GENOTYPING.
Status: ACCEPTED.
Implementation target date: PAG2016
Implemented by: Germinate, Cassavabase

### Markerprofiles by germplasmDbId [GET]
+ Response 200 (application/json)

        {
            "metadata" : {},
            "result" :  {
                germplasmDbId : 39393,
                markerProfiles : [
                    3939, 4484, 3993
                ]
            }
        }

# Group Germplasm attributes
Germplasm "attributes" are simply-inherited characteristics (characterization descriptors) that are not environment-dependent but
inherent in the germplasm line.  They include genes, QTLs, and genetic "traits" like wheat grain
hardness.  They are often evaluated by genotyping for diagnostic markers.
## List attributes available [/brapi/v1/attributes?attributeCategoryDbId=2]
Scope: OTHER. Status: IN DISCUSSION.
Implementation target date: PAG2016

### Attributes by attributeCategoryDbId [GET]
List available attributes.
###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|attributeCategoryDbId|string|internal database identifier, not stable |Y|
|code|string|abbreviation||
|uri|string|reference to external documentation, ontology etc.||
|name|string|for display, and stable identifier unique in the database|Y|
|description|string|||
|datatype|string|e.g. Categorical, Numeric, Boolean||
|values|string|array of all possible values for this attribute|Y|
+ Parameters
    + attributeCategoryDbId (optional, string, `2`) ... filter for kind of attributes
+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10000,
                    "totalPages": 1000
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data" : [
                    "attributeCategoryDbId": 1,
                    "code": "RHT",
                    "uri": "http://www.cropontology.org/rdf/CO_321:0000020",
                    "name": "Rht-B1b",
                    "description": "Allele of marker 11_4769, diagnostic for allele b of reduced-height gene Rht-B1",
                    "attributeCategoryId": 2,
                    "attributeCategoryName": "Agronomic",
                    "datatype": "Categorical",
                    "values": ["Present", "Absent", "Heterozygous"]
                ]
            }
        }

## List attribute categories [/brapi/v1/attributes/categories/?pageSize=10&page=2]
Scope: OTHER. Status: IN DISCUSSION.
Implementation target date: PAG2016

### Germplasm attribute categories [GET]
List all available attribute categories.
+ Parameters
    + pageSize (optional, number, `10000`) Number of attributes to return in one response
    + page (optional, number, `1`) ... Required if `pageSize` is given; and requires that `pageSize` be given. The first page is 1, not 0.

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data" : [
                    {
                        "attributeCategoryDbId": 1,
                        "name": "Morphological"
                    },
                    {
                        "attributeCategoryDbId": 2,
                        "name": "Agronomic"
                    }
                ]
            ]
        }

## Germplasm attribute values by germplasmDbId [/brapi/v1/germplasm/{germplasmDbId}/attributes?attributeList={attributeDbId},{attributeDbId}&pageSize=&page=]

### Germplasm attribute values [GET]
Values for all attributes by default.
+ Parameters
    + germplasmDbId (required, string, `993`) ... The germplasm characterized
    + attributeDbId (optional, string, `1`) ... Restrict the response to only the listed attributes.
    + pageSize (optional, number, `10000`) ... the number of attributes to return in one request, defaults to 1000.
    + page (optional, number, `1`) ... Required if `pageSize` is given; and requires that `pageSize` be given. The first page is 1, not 0.

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : [
                {
                    "germplasmDbId": "993",
                    "data": [
                        {
                            "attributeDbId": 1,
                            "attributeName": "Rht-B1b",
                            "attributeCode": "RHT",
                            "value": "Present",
                            "dateDetermined": "20070528"
                        }
                    ]
                }
            ]
        }


# Group Markers

## Markers Search [/brapi/v1/markers?name={name}&type={type}&matchMethod={matchMethod}&include={synonyms}&pageSize={pageSize}&page={page}]
Scope: CORE.  Status: ACCEPTED. Implemented by: Germinate

Other service requests use the server's internal `markerDbId`. This service returns marker records that provide the markerDbId.
For the requested name or synonym, returns an array (possibly empty) of marker records that match the search criteria.
- If there is none, an empty array is returned.
- If there is one or more than one match, returns an array of all marker records that match the search criteria.

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|result|object|data|Y|
|data|array of objects|Array (possibly empty) of marker records|Y|
|type|string|The type of marker, e.g. SNP|Y|
|markerDbId|string|Internal db identifier|Y|
|defaultDisplayName|string|A string representing the marker that will be meaningful to the user|Y|
|synonyms|array of string|List of other names for this marker||
|refAlt|array of string|List of the reference (as the first item) and alternatives (the remaining items)||
|analysisMethods|array of string|List of the genotyping platforms used to interrogate the marker||

### Search names to retrieve marker records [GET]
+ Parameters
    + name (optional, text, `11_10002 11_1% 11_1* 11_10?02`) ... The name or synonym.
    + matchMethod (optional, string, `wildcard`) ... Possible values are 'case_insensitive', 'exact'
    (case sensitive), 'wildcard' (which is case insensitive). Wildcard uses both '*' and '%' for any number of characters
    and '?' for one character matching. Default is exact.
    + include (optional, string, `synonyms`) ... Whether to include synonyms in the output.
    + type (optional, string, `SNP`) ... The type of the marker.
    + pageSize (optional, integer, `1000`) ... The size of the pages to be returned. Default is `1000`.
    + page (optional, integer, `10`) ... Which result page is requested
+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 1000,
                    "currentPage": 10,
                    "totalCount": 27338,
                    "totalPages": 28
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data": [
                    {
                        "markerDbId": "1185",
                        "defaultDisplayName": "11_10002",
                        "type": "SNP",
                        "synonyms": ["i_11_10002", "POPA1_0002"],
                        "refAlt": ["A", "T"],
                        "analysisMethods": ["illumina", "kasp"]
                    },
                    {
                        "markerDbId": "1186",
                        "defaultDisplayName": "11_11159",
                        "type": "SNP",
                        "synonyms": ["i_11_11159", "POPA1_1159"],
                        "refAlt": ["A", "T"],
                        "analysisMethods": ["illumina", "kasp"]
                    }
                ]
            }
        }

## Marker Details by markerDbId [/brapi/v1/markers/{markerDbId}]

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|metadata|object|pagination, status|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|
|status|list|code, message|Y|
|result|object|data|Y|
|data|array of objects|Array (possibly empty) of marker records|Y|
|type|string|The type of marker, e.g. SNP|Y|
|markerDbId|string|Internal db identifier|Y|
|defaultDisplayName|string|A string representing the marker that will be meaningful to the user|Y|
|synonyms|array of string|List of other names for this marker||
|refAlt|array of string|List of the reference (as the first item) and alternatives (the remaining items)||
|analysisMethods|array of string|List of the genotyping platforms used to interrogate the marker||

### Marker Details by id [GET]
+ Parameters
    + markerDbId (required, string, `1185`) ... the internal id of the marker

+ Response 200 (application/json)

        {
            "metadata": {
                "status": [],
                "datafiles": [],
                "pagination": {}
            },
            "result": {
                "markerDbId": "1185",
                "defaultDisplayName": "11_10002",
                "type": "SNP",
                "synonyms": ["i_11_10002", "POPA1_0002"],
                "refAlt": ["A", "T"],
                "analysisMethods": ["illumina", "kasp"]
            }
        }

# Group Markerprofiles

For the purposes of this API, the definition of markerprofile is *the allele calls for a specified germplasm line, for all markers, for a specified set of genotyping experiments or all experiments.*

Basic concepts in the **Breeding API**:

- *markerprofile*: A set of marker scores for a specific extract from a specific germplasm.
- *extract* : a preparation from a germplasm for an analysis.
- *sample*: an individual plant or tissue from a plant of a particular germplasm
- *germplasm*: a single genetic entity (cultivar, variety, accession, breeding line) used for analysis
- *marker*: a DNA sequence polymorphism, potentially localized to a single genomic location
- *allele*: one of the two possible states of a marker in each haploid chromosome complement of a specified germplasm, as determined in a specified experiment. A diploid organism has two alleles per marker.
- *missing*: a germplasm/marker/experiment combination for which no allele result is available, whether it was tested for or not

## Markerprofile search [/brapi/v1/markerprofiles?germplasm={germplasmDbId}&sample={sampleDbId}&extract={extractDbId}&method={methodDbId}&pageSize=100&page=4]
Scope: GENOTYPING.
Status: IN DISCUSSION.
Implementation target date: PAG2016

Implemented by: Germinate

For the requested Germplasm Id and/or Extract Id, returns the Markerprofile Id and number of non-missing allele calls
(marker/allele pairs).

### Retrieve Markerprofile Ids [GET]
+ Parameters
    + germplasmDbId (optional, string, `993`) ... The server's internal id for the Germplasm ID, as returned by the **Find markerprofile by Germplasm** service.
    + sampleDbId (optional, string, `184`) ... The server's internal id for the SampleDbId
    + extractDbId (optional, string, `84`) ... The server's internal id for the ExtractDbId
    + pageSize (optional, number, `10000`) ... The number of allele call results (marker/allele pairs) to be returned in the response. If multiple experiments are requested, some responses will contain the last results from one experiment followed by the first results from the next.
    + page (optional, number, `1`) ... Required if `pageSize` is given; and requires that `pageSize` be given. The first page is 1, not 0.
+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data" : [
                    {
                        "markerProfileDbId": "993",
                        "germplasmDbId" : "2374",
                        "sampleDbId" : "3937",
                        "extractDbId" : "3939",
                        "analysisMethod": "GoldenGate",
                        "resultCount": 1470
                    },
                    {
                        "markerProfileDbId": "994",
                        "germplasmDbId" : "2374",
                        "sampleDbId" : "1234",
                        "extractDbId" : "3939",
                        "analysisMethod": "GBS",
                        "resultCount": 1470
                    }
                ]
            }
        }

+ Response 400 (application/json)

        {
            "metadata" : {
                "pagination" : {},
                "status" : [
                    {
                        "code" : "ERR-1",
                        "message" : "Unsupported notation code."
                    }
                ],
                "datafiles": []
            },
            "result": []
        }

## Markerprofile data [/brapi/v1/markerprofiles/{germplasmDbId}?unknownString=&sepPhased=&sepUnphased=&expandHomozygotes=&pageSize=&page= ]
Scope: GENOTYPING.
Status: IN DISCUSSION.
Implementation target date: PAG2016

Implemented by: Germinate

For the requested markerprofile ID, returns the allele call for each marker.
[Example] (http://malt.pw.usda.gov/t3/wheatplus/brapi/v1/markerprofiles/1784_99/count?analysisMethod=GoldenGate)

**Allele encodings**

- Unknown data will by default be encoded by "N"
- Homozygotes are returned as a single occurance, e.g. AA -> A, GG -> G
- Phased heterozygotes are by default separated by a pipe ("|") and unphased heterozygotes are by default separated by a forward slash ("/")
- Dominant markers such as DArTs: 1 for present, 0 for absent

**Open issue:**
The pages of data will need to be sorted sensibly in order for the
requested page to be delivered consistently.  By map or genome position?
Alphabetically?
###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|markerprofileDbId|string|Unique in the database. Can be a catenation of unique IDs for germplasm and extract.|Y|
|germplasmDbId|string||Y|
|extractDbId|string||Y|
|analysisMethod|string|||
|data|object|array of marker-name/score pairs|Y|
|pagination|object|pageSize, currentPage, totalCount, totalPages|Y|

### Alleles By Markerprofile Id [GET]

+ Parameters
    + germplasmDbId (required, number, `993`) ... The server's internal id for the germplasm line, as returned by the **Find Germplasm by Name** service.
    + expandHomozygotes (optional, boolean, `false`) ... Should homozygotes NOT be collapsed into a single orrucance?
    + unknownString (optional, string, `-`) ... The string to use as a representation for missing data.
    + sepPhased (optional, string, `|`) ... The string to use as a separator for phased allele calls.
    + sepUnphased (optional, string, `/`) ... The string to use as a separator for unphased allele calls.
    + pageSize (optional, number, `10000`) ... The number of allele call results (marker/allele pairs) to be returned in the response. If multiple experiments are requested, some responses will contain the last results from one experiment followed by the first results from the next.
    + page (optional, number, `1`) ... Required if `pageSize` is given; and requires that `pageSize` be given. The first page is 1, not 0.

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },

            "result": {
                "germplasmDbId": 993,
                "extractDbId" : 38383,
                "markerprofileDbId": 37484,
                "analysisMethod": "GBS-Pst1",
                "data" : [ { "marker1": "A" }, { "marker2":"A/B" }, ... ]
           }
        }


## Markerprofile data [/brapi/v1/markerprofiles ]
Scope: GENOTYPING.
Status: IN DISCUSSION.
Implementation target date: PAG2016

## MarkerProfile Allele Matrix [/brapi/v1/allelematrix?unknownString=&sepPhased=&sepUnphased=&expandHomozygotes=&markerprofileDbId=100&markerprofileDbId=101&markerDbId=322&markerDbId=418&format=&pageSize=&page=]
Implemented by: Germinate

This uses a more efficient data structure and pagination for large number of markers.

### Scores through GET [GET]

Use GET when parameter size is less than 2K bytes.

+ Parameters
    + markerprofileDbId (required, string, `993`) ... The markerprofile db ids; for multiple, repeat the parameter.
    + markerDbId (optional, string, `322`) ... ids of the markers; if none are specified, results are returned for all markers in the database.
    + expandHomozygotes (optional, boolean, `false`) ... Should homozygotes NOT be collapsed into a single orrucance?
    + unknownString (optional, string, `-`) ... The string to use as a representation for missing data.
    + sepPhased (optional, string, `|`) ... The string to use as a separator for phased allele calls.
    + sepUnphased (optional, string, `/`) ... The string to use as a separator for unphased allele calls.
    + pageSize (optional, integer, `1000`) ... the number of allele calls reported per response page.
    + page (optional, integer, `10`) ... the requested response page

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 100,
                    "currentPage": 1,
                    "totalCount": 1,
                    "totalPages": 1
                },
                "status":[],
                "datafiles": []
            },
            "result" : {
                "makerprofileDbIds": ["markerprofileId1","markerprofileId2","markerprofileId3"],
                "data": [
                    { "markerId1":["A/B","A","A"] },
                    { "markerId2":["A","A/B","A"] },
                    { "markerId3":["A/B","A/B","B"] }
                ]
            }
        }


### Scores through POST [POST]

Use POST when parameter size is greater than 2K bytes.

- If no format is specified, this call returns the data in JSON form.
- If a format (other than JSON) is specified and the server supports this format, it will return the link to the exported data file in the "datafiles" field of the "metadata".
- If more than one format is requested at a time, the server will throw a "501 Not Implemented" error.

+ Parameters
    + markerprofileDbId (required, string, `993`) ... The markerprofile db ids; for multiple, repeat the parameter.
    + markerDbId (optional, string, `322`) ... ids of the markers; if none are specified, results are returned for all markers in the database.
    + expandHomozygotes (optional, boolean, `false`) ... Should homozygotes NOT be collapsed into a single orrucance?
    + unknownString (optional, string, `-`) ... The string to use as a representation for missing data.
    + sepPhased (optional, string, `|`) ... The string to use as a separator for phased allele calls.
    + sepUnphased (optional, string, `/`) ... The string to use as a separator for unphased allele calls.
    + format (optional, string, `tsv`) ... If specified, this indicates that the server should return the data in the specified data format. The link to the data file is then returned in the "datafiles" field of the "metadata". Initially only "tsv" is supported which returns the data in a tab-delimited format. MarkerprofileDbIds along the top, markerDbIds along the side and allele calls in the center.
    + pageSize (optional, integer, `1000`) ... the number of allele calls reported per response page.
    + page (optional, integer, `10`) ... the requested response page

+ Response 200 (application/json)

        {
            "metadata": {
                "status": [],
                "datafiles": [],
                "pagination": {
                    "pageSize": 100,
                    "currentPage": 1,
                    "totalCount": 1,
                    "totalPages": 1
                }
            },
            "result" : {
                "makerprofileDbIds": ["markerprofileId1","markerprofileId2","markerprofileId3"],
                "data" : [
                    { "markerId1":["A/B","A","A"] },
                    { "markerId2":["A","A/B","A"] },
                    { "markerId3":["A/B","A/B","B"] }
                ]
            }
        }

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 1,
                    "currentPage": 0,
                    "totalCount": 1,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": [
                    { "url": "https://my-fancy-server/files/allelematrix-1234.tsv" }
                ]
            },
            "result" : {
                "makerprofileDbIds": [],
                "data" : []
            }
        }



# Group Program
## Program list [/brapi/v1/programs?pageSize=&page=]
Call to retrieve a list of programs.

Implementation target date: PAG2016

### List programs [GET]

+ Parameters
    + pageSize (optional, integer, `23`) ... the number of programs to be reported per request
    + page (optional, integer, `292`) ... the response page requested

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data" : [
                    {
                        "programDbid": "123",
                        "name": "Wheat Resistance Program",
                        "abbreviation" : "DRP1",
                        "objective" : "Disease resistance",
                        "leadPerson" : "Dr. Henry Beachell"
                    },
                    {
                        "programDbId": "456",
                        "name": "Wheat Improvement Program",
                        "abbreviation" : "DRP2",
                        "objective" : "Yield improvement",
                        "leadPerson" : "Dr. Norman Borlaug"
                    }
                ]
            }
        }

# Group Trials

Services related to trials. Trials comprise of multiple studies.

## List trial Summaries [/brapi/v1/trials?locationDbId={locationDbId}&programDbId={programDbId}&pageSize={pageSize}&page={page}&active={active}&sortBy={sortBy}&sortOrder={sortOrder}]

Scope: CORE.
Status: IN DISCUSSION.
Implementation target date: PAG2016.

Get list of trials.

### List of trail summaries [GET]

+ Parameters
    + programDbId (optional, string, `1`) ... Program filter to only return trials associated with given program id.
    + locationDbId (optional, string, `212`) ... Filter by location
    + pageSize (optional, integer, `1000`)
    + page (optional, integer, `100`)
    + active (optional, boolean, `true/false`) ... Filter active status true/false.
    + sortBy (optional, boolean, `studyDbId`) ... Sort order. Name of the field to sorty by.
    + sortOrder (optional, boolean, `asc/desc`) ... Sort order direction. Ascending/Descending.

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 2,
                    "currentPage": 1,
                    "totalCount": 100,
                    "totalPages": 50
                },
            "status" : [],
            "datafiles": []
            },
            "result": {
                "data": [
                    {
                        "trialDbId" : 1,
                        "trialName" : "InternationalTrialA",
                        "programDbId": 27,
                        "programName": "International Yield Trial",
                        "startDate": "2007-06-01",
                        "endDate"  : "2008-12-31",
                        "active" : "true",
                        "studies" : [
                            {
                                "studyDbId" : 1,
                                "studyName" : "Zimbabwe Yield Trial",
                                "locationName" : "Zimbabwe"
                            },
                            {
                                "studyDbId" : 2,
                                "studyName" : "Kenya Yield Trial",
                                "locationName" : "Kenya"
                            }
                        ],
                        "additionalInfo" : {
                            "property1Name" : "property1Value",
                            "property2Name" : "property2Value",
                            "property3Name" : "property3Value"
                        }
                    }
                ]
            }
        }

## Get Trial By Id [/brapi/v1/trials/{trialDbId}]

Scope: CORE.
Status: IN DISCUSSION.
Implementation target date: PAG2016.

Get list of trials.

### Get trial by Id [GET]

+ Response 200 (application/json)

        {
            "metadata": null,
            "result": {

                "trialDbId" : 1,
                "trialName" : "InternationalTrialA",
                "programDbId": 27,
                "programName": "International Yield Trial",
                "startDate": "2007-06-01",
                "endDate"  : "2008-12-31",
                "active" : "true",
                "studies" : [
                    {
                        "studyDbId" : 1,
                        "studyName" : "Zimbabwe Yield Trial",
                        "locationName" : "Zimbabwe"
                    },
                    {
                        "studyDbId" : 2,
                        "studyName" : "Kenya Yield Trial",
                        "locationName" : "Kenya"
                    }
                ],
                "additionalInfo" : {
                    "property1Name" : "property1Value",
                    "property2Name" : "property2Value",
                    "property3Name" : "property3Value"
                }
            }
        }




# Group Study

Study is defined as a phenotyping experiment conducted at a single geographic location.


## Seasons [/brapi/v1/seasons?year=&pageSize=&page=]

Call to retrive all seasons (or years) in the database. (Added by Jan-Erik and Lukas 5/26/2016)

### List seasons or years [GET]
+ Parameters
    + year (optional, String, `2015`)
    + pageSize (optional, integer, `1`)
    + page (optional, integer, `1`)
+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data" : [
                    { "seasonDbId": "237", "season": "Fall", "year": "2015" },
                    { "seasonDbId": "238", "season": "Spring", "year": "2016" }
                ]
            }
        }

## Study Types [/brapi/v1/studyTypes?pageSize=&page=]

Call to retrieve the list of study types.

Implementation target date: PAG2016

### List study types [GET]
+ Parameters
    + pageSize (optional, integer, `1`)
    + page (optional, integer, `1`)

+ Response 200 (application/json)

        {
             "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data" : [
                    {
                        "name": "Crossing Nursery",
                        "description": "Description for Nursery study type"
                    },
                    {
                        "name": "Yield Trial",
                        "description": "Description for Nursery study type"
                    }
                ]
            }
        }

## List Study Summaries [/brapi/v1/studies?studyType={studyType}&seasonDbId={seasonDbId}&locationDbId={locationDbId}&programDbId={programDbId}&germplasmDbIds={germplasmDbIds}&observationVariableDbIds={observationVariableDbIds}&pageSize={pageSize}&page={page}&active={active}&sortBy={sortBy}&sortOrder={sortOrder}]
Scope: CORE.
Status: IN DISCUSSION.
Implementation target date: PAG2016. Implemented by: Germinate

Get list of studies

### List of study summaries [GET]

+ Parameters
    + studyType (optional, string, `Nursery`) ... Filter based on study type e.g. Nursery or Trial.
    + programDbId (optional, string, `1`) ... Program filter to only return studies associated with given program id.
    + locationDbId (optional, string, `212`) ... Filter by location
    + seasonDbId (optional, string, `2016E`) ... Filter by season or year
    + pageSize (optional, integer, `1000`)
    + page (optional, integer, `100`)
    + germplasmDbIds (optional, array, `["CML123", "CML"]`) ... Filter studies where specified germplasm have been used/tested
    + observationVariableDbIds (optional, array, `["CO-PH-123" , "Var-123"]`) ... Filter studies where specified observation variables have been measured
    + active (optional, boolean, `true/false`) ... Filter active status true/false.
    + sortBy (optional, boolean, `studyDbId`) ... Sort order. Name of the field to sorty by.
    + sortOrder (optional, boolean, `asc/desc`) ... Sort order direction. Ascending/Descending.

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 2,
                    "currentPage": 1,
                    "totalCount": 100,
                    "totalPages": 50
                },
                "status" : [],
                "datafiles": []
            },
            "result": {
                "data": [
                    {
                        "studyDbId": 35,
                        "name": "Earlygenerationtesting",
                        "trialDbId" : 1,
                        "trialName" : "InternationalTrialA",
                        "studyType": "Trial",
                        "seasons": ["2007 Spring", "2008 Fall"],
                        "locationDbId": 23,
                        "locationName": "Kenya",
                        "programDbId": 27,
                        "programName": "Drought Resistance Program A",
                        "startDate": "2007-06-01",
                        "endDate"  : "2008-12-31",
                        "studyType": "Trial",
                        "active" : "true",
                        "additionalInfo" : {
                            "property1Name" : "property1Value",
                            "property2Name" : "property2Value",
                            "property3Name" : "property3Value"
                        }
                    }
                    ,
                    {
                        "studyDbId": 345,
                        "name": "Earlygenerationtesting",
                        "trialDbId" : 1,
                        "trialName" : "InternationalTrialA",
                        "seasons": ["2007 Spring", "2008 Fall"],
                        "locationDbId": 33,
                        "locationName": "Zimbabwe",
                        "programDbId": 58,
                        "programName": "Drought Resistance Program B",
                        "startDate": "2005-06-01",
                        "endDate"  : "2008-12-31",
                        "studyType": "Trial",
                        "active" : "true",
                        "additionalInfo" : {
                            "property1Name" : "property1Value",
                            "property2Name" : "property2Value",
                            "property3Name" : "property3Value"
                        }
                    }
                ]
            }
        }

## Search Studies [/brapi/v1/studies]

Search studies

### Search studies [POST]

+ Request (application/json)

        {
           "studyType" : "Trial",
           "studyNames" : ["Study A", "StydyB"],
           "studyLocations" : ["Kenya", "Zimbabwe"],
           "programNames": ["Test Program", "Program2"],
           "germplasmDbIds" : [ "CML123", "CWL123"],
           "observationVariableDbIds": ["CO-PH-123", "Var-123"]
           "active" : "true",
           "sortBy" : "studyDbId",
           "sortOrder" : "desc"
        }

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 2,
                    "currentPage": 1,
                    "totalCount": 100,
                    "totalPages": 50
                },
                "status" : [],
                "datafiles": []
            },
            "result": {
                "data": [
                    {
                        "studyDbId": 35,
                        "name": "Earlygenerationtesting",
                        "trialDbId" : 1,
                        "trialName" : "InternationalTrialA",
                        "studyType": "Trial",
                        "seasons": ["2007 Spring", "2008 Fall"],
                        "locationDbId": 23,
                        "locationName": "Kenya",
                        "programDbId": 27,
                        "programName": "Drought Resistance Program A",
                        "startDate": "2007-06-01",
                        "endDate"  : "2008-12-31",
                        "studyType": "Trial",
                        "active" : "true",
                        "additionalInfo" : {
                            "property1Name" : "property1Value",
                            "property2Name" : "property2Value",
                            "property3Name" : "property3Value"
                        }
                    }
                    ,
                    {
                        "studyDbId": 345,
                        "name": "Earlygenerationtesting",
                        "trialDbId" : 1,
                        "trialName" : "InternationalTrialA",
                        "seasons": ["2005 Winter", "2006 Summer"],
                        "locationDbId": 33,
                        "locationName": "Zimbabwe",
                        "programDbId": 58,
                        "programName": "Drought Resistance Program B",
                        "startDate": "2005-06-01",
                        "endDate"  : "2008-12-31",
                        "studyType": "Trial",
                        "active" : "true",
                        "additionalInfo" : {
                            "property1Name" : "property1Value",
                            "property2Name" : "property2Value",
                            "property3Name" : "property3Value"
                        }
                    }
                ]
            }
        }

## Study Details [/brapi/v1/studies/{studyDbId}?expand={expand}]
Scope: CORE.
Status: IN DISCUSSION.
Implementation target date: PAG2016  Implemented by: Germinate

Notes: an optionalInfo field was added to provide a controlled vocabulary for less common data fields.

Retrieve the information of the study required for field data collection

###### Response data types

|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|studyDbId|Long|internal DB id ||
|studyName|string|Human readable name||
|years|list|list of years the trials is running||
|locationDbId|string| Location ID||
|optionalInfo|object| ||

### Retrieve study details [GET]

+ Parameters
    + studyDbId (required, string, `1`) ... Identifier of the study. Usually a number, could be alphanumeric.
    + expand (optional, string, `location`) ... Comma separated list of study details sections to expand. If none is specified, each details section will only produce a link to get details.

+ Response 200 (application/json)

        {
            "metadata" : {},
            "result" : {
                "studyDbId": 35,
                "studyName": "Earlygenerationtesting",
                "locationDbId": 34,
                "locationName" : "Zimbabwe",
                "seasons": ["2005 Winter", "2008 Summer"],
                "programDbId": 57,
                "optionalInfo" : {
                    "studyObjective" : "Increase yield",
                    "studyPUI" : "PUI string",
                    "studyType": "Trial",
                    "startDate": "2015-06-01",
                    "endDate": "2015-12-31"
                },
                "observationVariables" : "/brapi/v1/studies/{studyDbId}/observationVariables",
                "location" : "/brapi/v1/locations/{locationDbId}",
                "germplasm" : "/brapi/v1/studies/{studyDbId}/germplasm",
                "observations": "/brapi/v1/studies/{studyId}/observationunits",
                "layout" : "/brapi/v1/studies/{studyDbId}/layout"
            }
        }

## Study Observation Variables [/brapi/v1/studies/{studyDbId}/observationVariables]

List all the observation variables measured in the study.

### Retrieve study observation variables [GET]

+ Response 200 (application/json)

        {
            "metadata": {
                "status": [],
                "datafiles": [],
                "pagination": {
                    "pageSize": 1000,
                    "currentPage": 1,
                    "totalCount": 1,
                    "totalPages": 1
                }
            },
            "result": {
                "studyDbId": 123,
                "trialName": "myBestTrial",
                "data": [

                    {
                        "observationVariableId": "CO_334:0100622",
                        "observationVariableName": "carot_spectro",
                        "trait": "TraitName",
                        "method" : "Root to tip at maturity",
                        "scale" : "cm",
                        "dataType": "Numeric",
                        "scaleValidValues": {
                            "min": "0",
                            "max": "100",
                            "categories": null
                          }
                    },

                    {
                        "observationVariableId": "CO_334:112233",
                        "observationVariableName": "Flower Color",
                        "trait": "TraitName",
                        "method" : "Record manually",
                        "scale" : "Color",
                        "dataType": "Text",
                        "scaleValidValues": {
                            "min": null,
                            "max": null,
                            "categories": ["Red", "Blue", "Yellow"]
                          }
                    }

                ]
            }
        }


## Study Germplasm Details [/brapi/v1/studies/{studyDbId}/germplasm?pageSize={pageSize}&page={page}]

### Study Germplasm Details [GET]
+ Parameters
    + studyDbId (required, string, `1`) ... Identifier of the study. Usually a number, could be alphanumeric.
    + pageSize (optional, integer, `1000`) ... the number of germplasm entries to be returned in the response
    + page (optional, integer, `10`) ... the desired response page

+ Response 200 (application/json)

        {
            "metadata": {
                "status": [],
                "datafiles": [],
                "pagination": {
                    "pageSize": 1000,
                    "currentPage": 1,
                    "totalCount": 1,
                    "totalPages": 1
                }
            },
            "result": {
                "studyDbId": 123,
                "trialName": "myBestTrial",
                "data": [
                   {
                        "germplasmDbId": "382",
                        "entryNumber": "1",
                        "germplasmName": "Pahang",
                        "pedigree": "TOBA97/SW90.1057",
                        "seedSource": "SS1",
                        "accessionNumber": "ITC0609",
                        "germplasmPUI": "http://www.crop-diversity.org/mgis/accession/01BEL084609",
                        "synonyms": ["01BEL084609"]
                    },
                    {
                        "germplasmDbId": "394",
                        "entryNumber": "2",
                        "germplasmName": "Pahang",
                        "pedigree": "TOBA97/SW90.1057",
                        "seedSource": "SS2",
                        "accessionNumber": "ITC0727",
                        "germplasmPUI": "http://www.crop-diversity.org/mgis/accession/01BEL084727",
                        "synonyms": [ "01BEL084727"]
                    }
                ]
            }
        }


## Observation Levels [/brapi/v1/observationLevels]

Call to retrieve the list of supported observation levels.
Observation levels indicate the granularity level at which the measurements are taken.
The values are used to supply the `observationLevel` parameter in the "get observation units" call.

### List observation levels [GET]

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 2,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : ["plant", "plot"]
        }

## Field Data Collection [/brapi/v1/studies/{studyId}/observationunits?observationLevel={observationLevel}]

The main API call for field data collection, to retrieve all the observation units within a study.

### Get all observation units [GET]

+ Parameters
    + observationLevel (required, string, `plot`) ... The granularity level of observation units.
      Either `plotNumber` or `plantNumber` fields will be relavant depending on whether granularity is plot or plant respectively.

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 2,
                    "totalPages": 1
                },
                "status" : [],
                "datafiles": []
            },
            "result" : {
                "data" : [
                    {
                        "observationUnitDbId": "abc-123",
                        "observationUnitName": "Test-2016-location-34-575",
                        "germplasmDbId": "1",
                        "germplasmName": "IR-8",
                        "pedigree": "IR-8-FP/IR-8-MP",
                        "entryNumber": "1",
                        "entryType": "Test entry",
                        "plotNumber": "1",
                        "plantNumber" : "0",
                        "X" : "1",
                        "Y" : "1",
                        "replicate": "1",
                        "observations": [
                            {
                                "observationDbId": 153453453,
                                "observationVariableId": 18020,
                                "observationVariableName": "Plant_height",
                                "collector" : "Mr. Technician",
                                "observationTimeStamp" : "2015-06-16T00:53:26Z",
                                "value": null
                            },
                            {
                                "observationDbId": 23453454345,
                                "observationVariableId": 51496,
                                "observationVariableName": "GW100_g",
                                "collector" : "Mr. Technician",
                                "observationTimeStamp" : "2015-06-16T00:53:26Z",
                                "value": null
                            }
                        ]
                    },
                    {
                        "observatioUnitDbId": "abc-456",
                        "germplasmDbId": "2",
                        "germplasmName": "IR-9",
                        "pedigree": "IR-9-FP/IR-9-MP",
                        "enrtyNumber": "2",
                        "entryType": "Check entry",
                        "plotNumber": "2",
                        "plantNumber" : "0",
                        "X" : "1",
                        "Y" : "2",
                        "replicate": "1",
                        "observations": [
                            {
                                "observationDbId": 3,
                                "observationVariableId": 18020,
                                "observationVariableName": "Plant_height",
                                "collector" : "Mr. Technician",
                                "observationTimeStamp" : "2015-06-16T00:53:26Z",
                                "value": null
                            },
                            {
                                "observationDbId": 4,
                                "observationVariableId": 51496,
                                "observationVariableName": "GW100_g",
                                "collector" : "Mr. Technician",
                                "observationTimeStamp" : "2015-06-16T00:53:26Z",
                                "value": null
                            }
                        ]
                    }
                ]
            }
        }

### Save Or Update Measurements [POST]
Call to invoke for saving the measurements captured for all the observation units.

+ Parameters
    + observationLevel (required, string, `plot`) ... The granularity level of observation units.

+ Request (application/json)

        {
            "metadata" : null,
            "result" : {
                "transactionDbId": "83748382938",
                "commit" : "true/false",
                "data" : [
                    {
                        "observatioUnitDbId": "abc-123",
                        "observations": [
                            {
                                "observationDbId": 1,
                                "observationVariableId": 18020,
                                "observationVariableName": "Plant_height",
                                "collector" : "Mr. Technician",
                                "observationTimeStamp" : "2015-06-16T00:53:26Z",
                                "value": "11"
                            },
                            {
                                "observationDbId": 2,
                                "observationVariableId": 51496,
                                "observationVariableName": "GW100_g",
                                "collector" : "Mr. Technician",
                                "observationTimeStamp" : "2015-06-16T00:53:26Z",
                                "value": "111"
                            }
                        ]
                    },
                    {
                        "observatioUnitDbId": "abc-456",
                        "observations": [
                            {
                                "observationDbId": 3,
                                "observationVariableId": 18020,
                                "observationVariableName": "Plant_height",
                                "collector" : "Mr. Technician",
                                "observationTimeStamp" : "2015-06-16T00:53:26Z",
                                "value": "22"
                            },
                            {
                                "observationDbId": 4,
                                "observationVariableId": 51496,
                                "observationVariableName": "GW100_g",
                                "collector" : "Mr. Technician",
                                "observationTimeStamp" : "2015-06-16T00:53:26Z",
                                "value": "222"
                            }
                        ]
                    }
                ]
            }
        }

+ Response 200 (application/json)

+ Response 500 (application/json)

        {
            "metadata": {
                "status": [ {
                    "code": "Error",
                    "message": "Could not update observation values. Invalid data."
                }]
            }
        }

## Study Measurements as a Table [/brapi/v1/studies/{studyDbId}/table]
Scope: CORE.
Status: IN DISCUSSION. Implemented in Cassavabase, HIDAP and Germinate.
Notes:
Implementation target date: after PAG2016
Retrieve the details of the study required for field data collection. Includes actual trait data.

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|studyDbId|Long|internal DB id ||
|observationVariableDbId|Long | internal DB id for the traits measured ||
|observationVariableName|String| name of variable||
|data| object| List of lists, specifying the plotId, block, rep, germplasmId, and the phenotypic values||

### Retrieve study measurements as table [GET]

+ Parameters
    + studyDbId (required, string, `1`) ... Identifier of the study. Usually a number, could be alphanumeric.

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 1,
                    "currentPage": 1,
                    "totalCount": 1,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "headerRow": [ "year","studyDbId","studyName","locationDbId","locationName","germplasmDbId","germplasmName","observationUnitDbId","plotName","replicate","blockNumber", "observationTimestamp", "entryType", "X", "Y"],
                "observationVariableDbIds": [ "variable1Id", "variable2Id", "variable2Id"],  // for linking
                "observationVariableNames": [ "plant height", "fruit weight", "root weight" ],  // for display
                // the observationNames will follow the columns defined in the headerRow

                "data" :
                [
                    [1, 1, 1, "CIP1", ...],
                    [2, 1, 1, "CIP2", ...]
                ]
            }
        }


### Save study measurements as table [POST]

+ Request

        {
            "metadata": null,
            "result" : {
                "headerRow": [ "observationUnitDbId", "collector", "observationTimestamp", "variable1Id", "variable2Id", "variable3Id" ],
                "observationVariableDbIds": [ "variable1Id", "variable2Id", "variable2Id"],
                "data" :
                [
                    [1, "Mr. Technician1", "2015-06-16T10:00:00Z", "variable1Value", "variable2Value", "variable3Value"],
                    [2, "Mr. Technician2", "2015-06-16T11:00:01Z", "variable1Value", "variable2Value", "variable3Value"]
                ]
            }
        }

+ Response 200


## Plot Layout [/brapi/v1/studies/{studyDbId}/layout]
Scope: CORE.
Status: SUGGESTED.
Implementation target date: PAG2016
Notes: the plot information.

Retrieve the plot layout of the study with id {id}.

### Retrieve plot layout [GET]

+ Parameters
    + studyDbId (required, string, `1`) ... Identifier of the study. Usually a number, could be alphanumeric.

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination": {
                    "pageSize": 1,
                    "currentPage": 1,
                    "totalCount": 1,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result" : {
                "data" : [
                    {
                        "studyDbId": 1,
                        "observationUnitDbId": 11,
                        "observationUnitName": "ZIPA_68_Ibadan_2014",
                        "observationLevel": "plot",
                        "replicate": 1,
                        "germplasmDbId": 143,
                        "germplasmName": "ZIPA_68",
                        "blockNumber": 1,
                        "X": 20,
                        "Y": 22,
                        "entryType": "check/test/filler",
                        "additionalInfo" : {
                        }
                    }
                ]
            }
        }


## Get measurements by observation variable ids [/brapi/v1/studies/{studyDbId}/observations?observationVariableDbIds={observationVariableDbIds}]
Scope: CORE.
Status: SUGGESTED.
Implementation target date: PAG2016

Retrieve all plots where there are measurements for the given observation variables.

### Get measurements by observation variable ids [GET]

+ Parameters
    + studyDbId (required, string, `1`) ... Identifier of the study. Usually a number, could be alphanumeric.
    + observationVariableDbId (required, number, `2` ... Numeric `id` of that variable (combination of trait, unit and method)

+ Response 200 (application/json)

        {
            "metadata" :
                "status": [],
                "datafiles": [],
                "pagination": {
                    "pageSize": 1,
                    "currentPage": 1,
                    "totalCount": 1,
                    "totalPages": 1
                },
            "result" : {
                "data" : [
                    {
                        "studyDbId": 1,
                        "obserationDbId": 3383838,
                        "observationUnitDbId": 11,
                        "observationUnitName": "ZIPA_68_Ibadan_2014",
                        "observationLevel" : "plot",
                        "observationVariableDbId" : 393939,
                        "observationVariableName" : "Yield",
                        "observationTimestamp" : "2015-11-05 15:12",
                        "uploadedBy" : "dbUserId",
                        "operator" : "Jane Doe",
                        "germplasmDbId" : 8383,
                        "germplasmName": 143,
                        "value" : 5,
                    }
                ]
            }
        }



# Group Phenotypes

API to retrieve data (phenotype, environment variables) from studies. While the study calls focus on one study, calls in this section are for cross-study phenotypic data retrieval.

## Phenotype Search [/brapi/v1/phenotypes]
Scope: PHENOTYPING.
Status: ACCEPTED.
Implementation target: 2016. Implemented for GnpIS private data.
Use case: this section allows to get a dataset from multiple studies. It allows to integrate data from several databases.

###### Response data types
|Variable|Datatype|Description|Required|
|------|------|------|:-----:|
|observationUnitDbId|Long|internal DB id ||
|studyDbId|Long|internal DB id (Primary Key)||
|studyName | string |||
|studyLocationDbId| internal DB id (primary key) ||
|studyLocation|string| Location name or ID|Y|
|programName|string|||
|germplasmDbId|Long|internal DB id (Primary Key)||
|germplasmName|string| Display name for the germplasm. MCPD Name field||
|X,Y|string|||
|XLabel,YLabel|string|Example: X=row, Y = rootstock||
|observationVariableId|string| ID or PUI (DOI, URI, LSID)||
|observationVariableDbId | string | ||
|observations.season|string| Year or season, Phenotyping campain||
|value|string|||
|observationTimeStamp|string|ISO format "2006-07-03::10:00"||
|collector|string| Person or team who has made the observaiton||
|additionalInfo|array of key/value pair|||

### Phenotype Search [POST]

observationTimeStamp : Iso Standard
observationValue data type inferred from the ontology

+ Request (application/json)

        {
            "germplasmDbIds" : [ "Blabla", "Blabla2" ], // (optional, text, `986`) ... The name or synonym of external genebank accession identifiers
            "observationVariableDbIds" : [ 37373, 38338], // (optional, text, `CO_321:00000234`) ... The IDs of traits, could be ontology ID, database ID or PUI
            "studyDbIds" : [ 383, 2929, 3838383 ], // (optional, text, `2356`) ... The database ID / PK of the studies search parameter
            "locationDbIds" : [ 383838, 959595 ], // locations these traits were collected
            "programDbIds" : [ 3838, 33838 ], // list of programs that have phenotyped this trait
            "seasonDbIds" : [ 338, 3939 ], // (optional, text, `2001`) ... The year or Phenotyping campaign of a multiannual study (trees, grape, ...)
            "observationLevel" : "plot", // (optional, text, `plot`) ... The type of the observationUnit. Returns only the observaton unit of the specified type; the parent levels ID can be accessed through observationUnitStructure.
            "pageSize" : 100, // (optional, integer, `1000`) ... The size of the pages to be returned. Default is `1000`.
            "page" : 1, // (optional, integer, `10`) ... Which result page is requested
        }

+ Response 200 (application/json)

        {
            "metadata":
            {
                "pagination":
                {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 10,
                    "totalPages": 1
                },

                "status": [],
                "datafiles": []
            },

            "result":
            {
                "data":
                [   {
                        "observationDbId": 4884848,
                        "observationUnitDbId": 20,
                        "observationUnitName": "Montpellier-2016-xyz-blabla",
                        "studyDbId": 25,
                        "studyName": "Yield",
                        "studyLocationDbId" : 88484,
                        "studyLocation": "Montpellier",
                        "programName": "Inovine",
                        "observationLevel": "plot",
                        "germplasmDbId": 3425,
                        "germplasmName": "syrah",
                        "observationVariableId": "CO_321:0000045",
                        "observationVariableDbId" : 383838,
                        "season": "2005",
                        "value": "red",
                        "observationTimeStamp": null,
                        "collector": "John Doe and team",
                        "uploadedBy" : "Mr. Upload",

                        "additionalInfo" : {
                            "key1" : "value1",  // eg:
                            "studySet" : [ "National Network","Frost suceptibility network" ],
                            "studyPlatform": "Phenome",
                            "treatments":
                            [ {
                                "factor": "water regimen",
                                "modality": "water deficit"
                            } ],
                             "quality": "reliability of the observation",
                            "collectionFacilityLabel": "phenodyne",
                        },


                    },
                    { /// more entries...
                    },
                ],
            },
        }



# Group Traits

API to retrieve list of traits and their associated variables


## Traits [/brapi/v1/traits]
Scope: CORE.
Status: IN DISCUSSION.

Implemented by: Germinate

Call to retrieve a list of traits available in the system and their associated variables.

### List all traits [GET]
+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination" : {
                    "pageSize": 3,
                    "currentPage": 1,
                    "totalCount": 300,
                    "totalPages": 100
                }
                "status" : [],
                "datafiles": []
            },

            "result" : {
                "data" : [
                    {
                        "traitDbId": "123",
                        "traitId": "CO:123000007",
                        "name": "Plant Height",
                        "description": "Description of Plant Height",
                        "observationVariables": [
                            "CO_334:0100121",
                            "CO_334:0100122",
                            "CO_334:0100123"
                        ],
                        "defaultValue": null
                    },
                    {
                        "traitDbId": "123",
                        "traitId": "CO_334:0100620",
                        "name": "Carotenoid content",
                        "description": "Cassava storage root pulp carotenoid content",
                        "observationVariables": [
                            "CO_334:0100621",
                            "CO_334:0100622",
                            "CO_334:0100623"
                        ],
                        "defaultValue": null
                    }
                ],

            ]
        }



## Trait Detail [/brapi/v1/traits/{id}]
Scope: CORE.
Status: IN DISCUSSION.

Implemented by: Germinate

Retrieve the variables associated to a trait

+ Parameters
    + id (required, string, `464`) ... string containing id's of the traits

### Retrieve selected trait description [GET]

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {},
                "status": [],
                "datafiles": []
            },
            "result": {
                "traitDbId": "123",
                "traitId": "CO:123000007",
                "name": "Plant Height",
                "description": "Description of Plant Height",
                "observationVariables": [
                    "CO_334:0100121",
                    "CO_334:0100122",
                    "CO_334:0100123"
                ],
                "defaultValue": null
            }
        }

# Group Observation Variables

API to retrieve list and details of observation variables. An observation variable is composed by the unique combination of one Trait, one Method and one Scale.

## Data Types [/brapi/v1/variables/datatypes]

Call to retrieve a list of data types the variable can have.

### List Observation Variables data types [GET]
+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageSize": 10,
                    "currentPage": 1,
                    "totalCount": 6,
                    "totalPages": 1
                },
                "status": [],
                "datafiles": []
            },
            "result": ["Numeric", "Categorical", "Date", "Text", "Picture", "Boolean"]
        }

## Variables [/brapi/v1/variables]
Scope: CORE.
Status: ACCEPTED.

Call to retrieve a list of observationVariables available in the system.
`required` means the key has to be provided, but the value may be null.

### List all variables [GET]
+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination" : {
                    "pageSize": 3,
                    "currentPage": 1,
                    "totalCount": 300,
                    "totalPages": 100
                }
                "status" : [],
                "datafiles": []
            },

            "result" : {
                "data" : [
                            {
                                "observationVariableDbId": "123",    // required
                                "observationVariableOntologyId": "CO_334",  // required
                                "observationVariableOntologyName": "Cassava",  //required
                                "observationVariableId": "CO_334:0100622",  // required
                                "observationVariableName": "caro_spectro",   // required
                                "observationVariableSynonyms": ["Carotenoid content by spectro", ""],
                                "observationVariableContextOfUse": ["", ""],
                                "observationVariableGrowthStage": "",
                                "observationVariableStatus": "recommended",
                                "observationVariableXref": "",
                                "observationVariableInstitution": "",
                                "observationVariableScientist": "",
                                "observationVariableDate": "",
                                "observationVariableLanguage": "EN",
                                "observationVariableCrop": "Cassava",
                                "trait": {                               // required
                                    "traitDbId": "123",   // required
                                    "traitId": "CO_334:0100620",   // required
                                    "traitName": "Carotenoid content",   // required
                                    "traitDescription": "Cassava storage root pulp carotenoid content",
                                    "traitClass": "",
                                    "traitSynonyms": ["", ""],
                                    "traitMainAbbreviation": "",
                                    "traitAlternativeAbbreviations": ["", ""],
                                    "traitEntity": "root",
                                    "traitAttribute": "carotenoid",
                                    "traitStatus": "recommended",
                                    "traitXref": ""
                                },
                                "measurementMethod": {   // required
                                    "methodId": "CO_334:0010320",
                                    "methodName": "Visual Rating:total carotenoid by chart_method",
                                    "methodDescription": "Assessment of the level of yellowness in cassava storage root pulp using the tc chart",
                                    "methodClass": "",
                                    "methodFormula": "",
                                    "methodReference": ""
                                },
                                "scale": {    // required
                                    "scaleId": "CO_334:0100526",
                                    "scaleName": "ug/g",
                                    "scaleClass": "",
                                    "dataType" : "numeric",
                                    "scaleDecimalPlaces": 2,
                                    "scaleXref": "",
                                    "scaleValidValues": {
                                        "min": "0",
                                        "max": "100",
                                        "categories": null
                                    }
                                },
                                "defaultValue": null    // required
                            }
                        ]
            }
        }



## Variables by category [/brapi/v1/variables/category/{categoryName}]
Scope: CORE.
Status: IN DISCUSSION.

Call to retrieve a list of variables for a given category as defined in the crop ontology (Phenology, stress, etc.).
+ Parameters
    + categoryName (required, string, `Phenology`) ... string containing the name of the category

### List all variables [GET]
+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination" : {
                    "pageSize": 3,
                    "currentPage": 1,
                    "totalCount": 300,
                    "totalPages": 100
                }
                "status" : [],
                "datafiles": []
            },

            "result" : {
                "data" : [
                   {
                        "observationVariableDbId": "123",
                        "observationVariableId": "CO_334:0100622",
                        "name": "caro_spectro",
                        "fullName": "Carotenoid content by spectro",
                        "trait": {
                            "traitDbId": "123",
                            "traitId": "CO_334:0100620",
                            "name": "Carotenoid content",
                            "description": "Cassava storage root pulp carotenoid content"
                        },
                        "measurementMethod": {
                            "methodId": "CO_334:0010320",
                            "name": "Visual Rating:total carotenoid by chart_method",
                            "description": "Assessment of the level of yellowness in cassava storage root pulp using the tc chart"
                        },
                        "scale": {
                            "scaleId": "CO_334:0100526",
                            "name": "ug/g",
                            "dataType": "Numeric",
                            "validValues": {
                                "min": "0",
                                "max": "100",
                                "categories": null
                            }
                        },
                        "defaultValue": null
                    },
                    {
                        "observationVariableDbId": "124",
                        "observationVariableId": "CO_334:0100621",
                        "name": "Carot_chart",
                        "fullName": "Carotenoid content by chart",
                        "trait": {
                            "traitDbId": "124",
                            "traitId": "CO_334:0100620",
                            "name": "Carotenoid content",
                            "description": "Cassava storage root pulp carotenoid content"
                        },
                        "measurementMethod": {
                            "methodId": "CO_334:0010319",
                            "name": "Visual Rating:total carotenoid by chart_method",
                            "description": "Assessment of the level of yellowness in cassava storage root pulp using the tc chart",
                        },
                        "scale": {
                            "scaleId": "CO_334:0100525",
                            "name": "8pt scale",
                            "dataType": "Categorical",
                            "validValues": {
                                "min": null,
                                "max": null,
                                "categories": [
                                    "white",
                                    "light cream",
                                    "cream",
                                    "light yellow",
                                    "yellow",
                                    "deep yellow",
                                    "orange",
                                    "pink"
                                ]
                            }
                        },
                        "defaultValue": null
                    }
                ]
            }
        }

## Variable Detail [/brapi/v1/variables/{id}]
Scope: CORE.
Status: ACCEPTED.

Retrieve variable details

+ Parameters
    + id (required, string, `CO_334:0100622`) ... string containing id's of the variables

### Retrieve selected variable description [GET]

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": null,
                "status": [],
                "datafiles": []
            },
            "result": {
                "traitDbId": "123",
                "observationVariableId": "CO_334:0100622",
                "name": "carot_spectro",
                "fullName": "Carotenoid content by spectro",
                "trait": {
                    "traitId": "CO_334:0100620",
                    "name": "Estimation :Total Carotenoid Content_method",
                    "description": "Total extracted carotenoids in cassava storage root as estimated by spectrophotometer"
                },
                "measurementMethod": {
                    "methodId": "CO_334:0010320",
                    "name": "Visual Rating:total carotenoid by chart_method",
                    "description": "Assessment of the level of yellowness in cassava storage root pulp using the tc chart"
                },
                "scale": {
                    "scaleId": "CO_334:0100526",
                    "name": "ug/g",
                    "dataType": "Numeric",
                    "validValues": {
                        "min": "0",
                        "max": "100",
                        "categories": null
                    }
                },
                "defaultValue": null
            }
        }


# Group Genome Maps
Retrieving genetic or physical maps
- type: [Genetic | Physical]
- unit: [cM | Mb]
- linkageGroup: may be scaffold (i.e. linkageGroupCount may consist of chromosomes or scaffolds or a combination of the two)
- For genetic maps, map naming convention should tell whether it is a consensus or mapping population (name of population) map for genetic maps
- For physical maps, map naming convention should tell whether it is a reference genome (name of line) or pan-genome

## Genome Map  [/brapi/v1/maps?species={speciesId}&pageSize={pageSize}&page={page}&type={type}]

Implemented by: Germinate

Get list of maps

do we need list of parents and specify mapping population?

### Get list of maps [GET]

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination" : {
                    "pageSize": 30,
                    "currentPage": 2,
                    "totalCount": 40,
                    "totalPages": 2
                }
                "status" : [],
                "datafiles": []
            },
            "result": {
                "data" : [
                    {
                        "mapDbId": 1,
                        "name": "Some Map",
                        "species": "Some species",
                        "type": "Genetic",
                        "unit": "cM",
                        "publishedDate": "2008-04-16",
                        "markerCount": 1000,
                        "linkageGroupCount": 7,
                        "comments": "This map contains ..."
                    },
                    {
                        "mapDbId": 2,
                        "name": "Some Other map",
                        "species": "Some Species",
                        "type": "Genetic",
                        "unit": "cM",
                        "publishedDate": "2009-01-12",
                        "markerCount": 1501,
                        "linkageGroupCount": 7,
                        "comments": "this is blah blah"
                    }
                ]
            }
        }

## Genome Map Details [/brapi/v1/maps/{mapDbId}]

Implemented by: Germinate

### Get map details [GET]

Provides the number of markers on each linkageGroup and the max position on the linkageGroup

+ Paramters
    + mapID (required)
    + Should we provide pagination?

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination" : {
                    "pageSize": 30,
                    "currentPage": 2,
                    "totalCount": 40,
                    "totalPages": 2
                }
                "status" : [],
                "datafiles": []
            },
            "result": {
                "mapDbId": "id",
                "name": "Some map",
                "type": "Genetic",
                "unit": "cM",
                "linkageGroups": [
                    {
                        "linkageGroupId": 1,
                        "numberMarkers": 100000,
                        "maxPosition": 10000000
                    },
                    {
                        "linkageGroupId": 2,
                        "numberMarkers": 1247,
                        "maxPostion": 12347889
                    }
                ]
            }
        }

## Genome Map Data [/brapi/v1/maps/{mapDbId}/positions?linkageGroupIdList={linkageGroupId,linkageGroupId}&pageSize={pageSize}&pageNumber={pageNumber}]

Implemented by: Germinate

### Get map data [GET]

markers ordered by linkageGroup and position

+ Parameters
   + mapDbId (required, string, `6`) ... unique id of the map
   + linkageGroupIdList (optional, comma separated list of strings)

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination" : { "pageSize": 30, "currentPage": 2, "totalCount": 40, "totalPages":2 },
                "status": [],
                "datafiles": []
            },
            "result": {
                "data" : [
                    {
                        "markerDbId": 1,
                        "markerName": "marker1",
                        "location": "1000",
                        "linkageGroup": "1A"
                    }, {
                        "markerDbId": 2,
                        "markerName": "marker2",
                        "location": "1001",
                        "linkageGroup": "1A"
                    }
                ]
            }
        }
## Genome Map Data by range on linkageGroup [/brapi/v1/maps/{mapDbId}/positions/{linkageGroupId}?min={min}&max={max}&pageSize={pageSize}&pageNumber={pageNumber}]

### Get map data by range on linkageGroup [GET]

markers ordered by linkageGroup and position

+ Parameters
   + mapDbId (required, string, `6`) ... unique id of the map
   + linkageGroupId (required, string)
   + min (optional) ... minimum position on linkage group
   + max (optional) ... maximumn position on linkage group

+ Response 200 (application/json)

        {
            "metadata" : {
                "pagination" : { "pageSize": 30, "currentPage": 2, "totalCount": 40, "totalPages":2 },
                "status" : [],
                "datafiles": []
            }
            "result": {
                "data" : [
                    {
                        "markerDbId": 1,
                        "markerName": "marker1",
                        "location": "1000"
                    }, {
                        "markerDbId": 2,
                        "markerName": "marker2",
                        "location": "1001"
                    }
                ]
            }
        }


# Group Locations

## List [/brapi/v1/locations?locationType={locationType}]

Implemented by: Germinate

Get a list of locations.

* The `countryCode` is as per [ISO_3166-1_alpha-3](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3) spec.
* `altitude` is in meters.

**Note**: Consider revising to describe polygon lat/lan points and check if adopting http://geojson.org/ is worth doing for v1.

### List locations [GET]

+ Parameters
    + locationType (optional, string, `Breeding Locations`) - Filter by location type specified.

+ Response 200 (application/json)

        {
            "metadata": {
                "pagination": {
                    "pageNumber": 1,
                    "pageSize": 2,
                    "totalCount": 100,
                    "totalPages": 50
                },
                "status" : [],
                "datafiles": []
            },
            "result": {
                "data" : [
                    {
                        "locationDbId": 1,
                        "locationType" : "Breeding Location",
                        "name": "Ibadan",
                        "abbreviation": "IB",
                        "countryCode": "NGA",
                        "countryName": "Nigeria",
                        "latitude": -21.5,
                        "longitude": 165.5,
                        "altitude": 12,
                        "additionalInfo":
                        {
                            "AnnualMeanRain" : "value",
                            "SoilDescription" :"23"
                        }
                    },
                    {
                        "locationDbId": 2,
                        "locationType" : "Storage Location",
                        "name": "Goa",
                        "abbreviation": "GO",
                        "countryCode": "IND",
                        "countryName": "India",
                        "latitude": 28.36,
                        "longitude": 77.12,
                        "altitude": 10,
                        "additionalInfo": {
                            "name1" : "value1",
                            "name2" : "value2"
                        }
                    }
                ]
            }
        }



# Group Samples

API methods for traking/managing plant samples and realted metatdata. Sample in the context of these BrAPI calls, is defined as the actual bilogical plant material collected from the field.
For more context on a possible user story see the [GOBII BMS Usecase diagram](http://www.gliffy.com/go/publish/image/10906911/L.png).


## Take a Sample [/brapi/v1/samples]

### Take a sample [PUT]

Call to register the event of a sample being taken. Sample ID is assigned as a result of the operation and returned in response.

+ Request (application/json)

        {
          "plotId": "PlotId-123",
          "plantId" : "PlantID-123",
          "takenBy": "Mr. Technician",
          "sampleDate": "2016-07-27",
          "sampleType" : "TypeOfSample",
          "tissueType" : "TypeOfTissue",
          "notes": "Cut from infected leaf"
        }

+ Response 200

        {
            "metadata": null,
            "result": {
                "sampleId" : "Unique-Plant-SampleID"
            }
        }

## Retrieve sample metadata [/brapi/v1/samples/{sampleId}]

### Retrieve sample metadata [GET]

+ Response 200 (application/json)

        {
            "metadata" : null,
            "result": {
              "studyDbId": "StudyId-123",
              "locationDbId": "LocationId-123",
              "plotId": "PlotId-123",
              "plantId" : "PlantID-123",
              "sampleId" : "Unique-Plant-SampleID",
              "takenBy": "Mr. Technician",
              "sampleDate": "2016-07-27",
              "sampleType" : "TypeOfSample",
              "tissueType" : "TypeOfTissue",
              "notes": "Cut from infected leaf",
              "studyName": "Yield Trial A",
              "season": "Summer",
              "locationName": "Kenya",
              "entryNumber": 11,
              "plotNumber": 1,
              "germplasmDbId": 15,
              "plantingDate": "20160101",
              "harvestDate": "20160630"
            }
        }

